ompi/mca/osc/ucx/osc_ucx_active_target.c

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2017. ALL RIGHTS RESERVED.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+
+#include "ompi/mca/osc/osc.h"
+#include "ompi/mca/osc/base/base.h"
+#include "ompi/mca/osc/base/osc_base_obj_convert.h"
+
+#include "osc_ucx.h"
+
+
+int ompi_osc_ucx_fence(int assert, struct ompi_win_t *win) {
+    ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t*) win->w_osc_module;
+    int ret = OMPI_SUCCESS;
+    ucs_status_t status;
+
+    if (module->epoch_type.access != NONE_EPOCH &&
+        module->epoch_type.access != FENCE_EPOCH) {
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    if (assert & MPI_MODE_NOSUCCEED) {
+        module->epoch_type.access = NONE_EPOCH;
+    } else {
+        module->epoch_type.access = FENCE_EPOCH;
+    }
+
+    if (assert & MPI_MODE_NOPRECEDE) {
+        ret = module->comm->c_coll->coll_barrier(module->comm,
+                                                 module->comm->c_coll->coll_barrier_module);
+        return ret;
+    }
+
+    status = ucp_worker_flush(mca_osc_ucx_component.ucp_worker);
+    if (status != UCS_OK) {
+        opal_output_verbose(1, ompi_osc_base_framework.framework_output,
+                            "%s:%d: ucp_worker_flush failed: %d\n",
+                            __FILE__, __LINE__, status);
+        return OMPI_ERROR;
+    }
+
+    module->global_ops_num = 0;
+    memset(module->per_target_ops_nums, 0,
+           sizeof(int) * ompi_comm_size(module->comm));
+
+    ret = module->comm->c_coll->coll_barrier(module->comm,
+                                             module->comm->c_coll->coll_barrier_module);
+    return ret;
+}
+
+int ompi_osc_ucx_start(struct ompi_group_t *group, int assert, struct ompi_win_t *win) {
+    ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t*) win->w_osc_module;
+
+    if (module->epoch_type.access != NONE_EPOCH &&
+        module->epoch_type.access != FENCE_EPOCH) {
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    module->epoch_type.access = START_COMPLETE_EPOCH;
+
+    OBJ_RETAIN(group);
+    module->start_group = group;
+
+    if ((assert & MPI_MODE_NOCHECK) == 0) {
+        int size = ompi_group_size(module->start_group);
+
+        while (module->state.post_count != (uint64_t)size) {
+            /* not sure if this is required */
+            ucp_worker_progress(mca_osc_ucx_component.ucp_worker);
+        }
+
+        module->state.post_count = 0;
+
```

Comment:
@hjelmn @jladd-mlnx I tested with ompi-tests/onesided and ompi-tests/mpi_test_suite/one-sided. I am trying to trigger MTT to test my branch, so that we can run entire ompi-tests.